tests/test_cases/test_cocotb/test_cocotb.py

```diff
@@ -1466,3 +1466,48 @@ def test_undecorated_coroutine_yield(dut):
     global last_ordered_test
     val, last_ordered_test = last_ordered_test, 1
     assert val == 2
+
+
+@cocotb.test()
+async def test_trigger_lock(dut):
+    """
+    Simple test that checks to see if context management is kept. The
+    resource value is checked at certain points if it equals the expected
+    amount, which is easily predictable if the context management is working.
+    """
+    resource = 0
+    lock = Lock()
+
+    async def co():
+        nonlocal resource
+        await Timer(10, "ns")
+        async with lock:
+            for i in range(4):
+                resource += 1
+                await Timer(10, "ns")
+
+    cocotb.fork(co())
+    async with lock:
+        for i in range(4):
+            resource += 1
+            await Timer(10, "ns")
+    assert resource == 4
```

Comment:
How can you guarantee this is `4` here? Couldn't the other coroutine have grabbed the lock and incremented already?

Reply:
Right now this test is assuming that releasing a lock is not a yield point. I don't know if it's necessary to guarantee that, so I'd prefer not to assume it in the test.

Reply:
That question opens a can of worms... Ideally, no. We do not yield control back to the scheduler between the final await in the with block (still holding the lock) and the next await, which is after this line. **And the `release()` method does not cause and immediate context switch or re-enter the scheduler**.

Lines in bold I'm uncertain about. Lines in bold *should* hold true. I hope this thing doesn't do that...

Reply:
You're correct that we don't re-enter the scheduler today, but it's not clear to me that this is a promise we need to make.

I'd be more comfortable if we change this test to the following, simply because then the test would still pass if we swapped from coroutines to threads.
```suggestion
        assert resource == 4
```

Reply:
> How can you guarantee this is `4` here? Couldn't the other coroutine have grabbed the lock and incremented already?

Hmmm. So, I was making an assumption that releasing the lock doesn't trigger the scheduler. I changed the test so that it should be guaranteed to work.

Reply:
I think it's a promise we should make. It makes concurrency *much* easier to reason about when you know that context switches only happen when you see syntax for it. Since you've mentioned Trio, I actually looked into it the other day, and they make a BIG DEAL about that. They call them "checkpoints".

Reply:
Right, but the whole point in `async with` is it has the potential for _two_ checkpoints, one at the start and one at the end. If you want a one-sided checkpoint, you can use `with (await lock)` instead of `async with lock`.

Reply:
From their docs:

> Partial exception for async context managers: Both the entry and exit of an async with block are defined as async functions; but for a particular type of async context manager, it’s often the case that only one of them is able to block, which means only that one will act as a checkpoint. This is documented on a case-by-case basis.

So basically, entrypoints are a good idea, but you can't see them from syntax alone.